conformance/utils/kubernetes/apply.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/ioutil"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MustApplyWithCleanup creates or updates Kubernetes resources defined with the
+// provided YAML file and registers a cleanup function for resources it created.
+// Note that this does not remove resources that already existed in the cluster.
+func MustApplyWithCleanup(t *testing.T, c client.Client, path string, gcName string, cleanup bool) {
+	b, err := ioutil.ReadFile(path)
+	require.NoErrorf(t, err, "error reading %s file", path)
+
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(b), 4096)
+	for {
+		uObj := unstructured.Unstructured{}
+		if decodeErr := decoder.Decode(&uObj); decodeErr != nil {
+			if errors.Is(decodeErr, io.EOF) {
+				break
+			}
+			t.Logf("manifest: %s", string(b))
+			require.NoErrorf(t, decodeErr, "error parsing manifest")
+		}
+		if len(uObj.Object) == 0 {
+			continue
+		}
+
+		if uObj.GetKind() == "Gateway" {
+			err = unstructured.SetNestedField(uObj.Object, gcName, "spec", "gatewayClassName")
+			require.NoErrorf(t, err, "error setting GatewayClass")
+		}
+		namespacedName := types.NamespacedName{Namespace: uObj.GetNamespace(), Name: uObj.GetName()}
+		fetchedObj := uObj.DeepCopy()
+		err := c.Get(context.TODO(), namespacedName, fetchedObj)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				require.NoErrorf(t, err, "error getting resource")
+			}
+			t.Logf("Creating %s %s", uObj.GetName(), uObj.GetKind())
+			err = c.Create(context.TODO(), &uObj)
```

Comment:
Please replace all contexts with a 10 or 15-second timeout to avoid long-hanging tests. We can fine-tune the timeout later on.
A better (more tedious) approach would be to have a context being propagated up from the top. Either is fine, please avoid requests without timeout.

Reply:
Good call, I've updated to use context more intentionally throughout, but will work on a larger follow up in the next PR to see where it makes sense to propagate context further.